imports/zk_goerli_eth.leo

```diff
@@ -0,0 +1,151 @@
+    }
+
+    struct AllowPair {
+        holder: address,
+        spender: address,
+    }
+
+    /* permission */
+    transition initial(holder: address) {
+        return then finalize (holder);
+    }
+
+    finalize initial(holder: address) {
+        let is_initial: bool = Mapping::get_or_use(initialed, true, false);
+        assert_eq(is_initial, false);
+
+        Mapping::set(contract_ownership, true, holder);
+    }
+
+    transition transfer_ownership (recipient: address) {
+        return then finalize (recipient, self.caller);
+    }
+
+    finalize transfer_ownership (recipient: address, caller: address) {
+        let is_initial: bool = Mapping::get_or_use(initialed, true, false);
+        assert_eq(is_initial, true);
+
+        let current_owner: address = Mapping::get(contract_ownership, true);
+        assert_eq(current_owner, caller);
+
+        Mapping::set(contract_ownership, true, recipient);
+    }
+
+    /* mint & burn */
+
+    transition mint_public(public receiver: address, public amount: u128) {
+        return then finalize(receiver, amount, self.caller);
+    }
+
+    finalize mint_public(public receiver: address, public amount: u128, public caller: address) {
+        let current_owner: address = Mapping::get(contract_ownership, true);
+        assert_eq(current_owner, caller);
+        let current_amount: u128 = Mapping::get_or_use(account, receiver, 0u128);
+        Mapping::set(account, receiver, current_amount + amount);
+    }
+
+    transition mint_private(receiver: address, amount: u128) -> token {
+        return token {
+            owner: receiver,
+            amount: amount,
+        } then finalize (self.caller);
+    }
+
+    finalize mint_private (public caller: address) {
+        let current_owner: address = Mapping::get(contract_ownership, true);
+        assert_eq(current_owner, caller);
+    }
+
+    transition burn_public (public receiver: address, public amount: u128) {
+        return then finalize(receiver, amount, self.caller);
+    }
+
+    finalize burn_public (public receiver: address, public amount: u128, public caller: address) {
+        let current_owner: address = Mapping::get(contract_ownership, true);
+        assert_eq(current_owner, caller);
+
+        let current_amount: u128 = Mapping::get_or_use(account, receiver, 0u128);
+        Mapping::set(account, receiver, current_amount - amount);
+    }
+
+    transition burn_private (receiver: address, note: token, amount: u128) -> token {
+        return token {
+            owner: note.owner,
+            amount: note.amount - amount,
+        } then finalize (self.caller);
+    }
+
+    finalize burn_private (public caller: address) {
+        let current_owner: address = Mapping::get(contract_ownership, true);
+        assert_eq(current_owner, caller);
+    }
+
+    /* Transfer */
+    transition transfer_public(public receiver: address, public amount: u128) {
+        return then finalize(self.signer, receiver, amount);
```

Comment:
Here `self.caller` makes more sense for me than `self.signer`, what do you think?

Reply:
Because this situation:
User  --Call--> A Contract --Call--> this Contract's transfer_public
if we use self.caller, it will cost the A Contract's balance.

Reply:
That's the same as ERC20, which is fine IMO.

Reply:
I got it. fix then.

Reply:
Not pushed?

Reply:
pushed